tower/src/retry/backoff.rs

```diff
@@ -0,0 +1,228 @@
+        }
+        if jitter < 0.0 {
+            return Err(InvalidBackoff("jitter must not be negative"));
+        }
+        if jitter > 100.0 {
+            return Err(InvalidBackoff("jitter must not be greater than 100"));
+        }
+        if !jitter.is_finite() {
+            return Err(InvalidBackoff("jitter must be finite"));
+        }
+
+        Ok(ExponentialBackoff {
+            min,
+            max,
+            jitter,
+            rng,
+            iterations: 0,
+        })
+    }
+
+    fn base(&self) -> time::Duration {
+        debug_assert!(
+            self.min <= self.max,
+            "maximum backoff must not be less than minimum backoff"
+        );
+        debug_assert!(
+            self.max > time::Duration::from_millis(0),
+            "Maximum backoff must be non-zero"
+        );
+        self.min
+            .checked_mul(2_u32.saturating_pow(self.iterations))
+            .unwrap_or(self.max)
+            .min(self.max)
+    }
+
+    /// Returns a random, uniform duration on `[0, base*self.jitter]` no greater
+    /// than `self.max`.
+    fn jitter(&mut self, base: time::Duration) -> time::Duration {
+        if self.jitter == 0.0 {
+            time::Duration::default()
+        } else {
+            let jitter_factor = self.rng.gen::<f64>();
+            debug_assert!(
+                jitter_factor > 0.0,
+                "rng returns values between 0.0 and 1.0"
+            );
+            let rand_jitter = jitter_factor * self.jitter;
+            let secs = (base.as_secs() as f64) * rand_jitter;
+            let nanos = (base.subsec_nanos() as f64) * rand_jitter;
+            let remaining = self.max - base;
+            time::Duration::new(secs as u64, nanos as u32).min(remaining)
+        }
+    }
+}
+
+impl Backoff for ExponentialBackoff {
+    type Future = tokio::time::Sleep;
+
+    fn next_backoff(&mut self) -> Self::Future {
+        let base = self.base();
+        let next = base + self.jitter(base);
+
+        self.iterations += 1;
+
+        tokio::time::sleep(next)
+    }
+}
+
+impl<R> Clone for ExponentialBackoff<R>
+where
+    R: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            min: self.min,
+            max: self.max,
+            jitter: self.jitter,
+            rng: self.rng.clone(),
+            iterations: 0,
+        }
+    }
+}
+
+impl Default for ExponentialBackoff {
+    fn default() -> Self {
+        let rng = SmallRng::from_rng(&mut thread_rng()).expect("RNG must be valid");
+        ExponentialBackoff::new(
+            Duration::from_millis(50),
+            Duration::from_millis(u64::MAX),
+            0.99,
+            rng,
+        )
+        .expect("Unable to create ExponentialBackoff")
+    }
+}
+
+/// Backoff validation error.
+#[derive(Clone, Debug)]
+pub struct InvalidBackoff(&'static str);
+
+impl Display for InvalidBackoff {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
```

Comment:
hmm, i don't think the `fmt::Display` implementation should just call the `fmt::Debug` implementation? maybe something like this:
```suggestion
        write!(f, "invalid backoff: {}", self.0)
```